File: packages/legacy-import/src/secret-policy.ts
// M11 secret policy (clean-room). Provider/model API secrets are SKIP by default;
// explicit consent moves them to a SecretStore as credential_ref only. Seller
// cookies/session tokens/passwords/auth state are NEVER imported.
import type { LegacyImportOptions } from "./types.js";

export const SELLER_SECRET_FIELDS = [
  "cookie", "cookies", "session", "session_token", "token", "password", "passwd",
  "auth", "authorization", "device_fingerprint", "license", "login_state", "credential",
];

export const PROVIDER_SECRET_FIELDS = [
  "api_key", "apiKey", "apikey", "secret", "secret_key", "access_key", "app_secret",
];

export interface SecretDetection {
  secret_fields_detected: string[];
  seller_secrets_detected: string[];
  provider_secrets_detected: string[];
  plaintext_must_not_persist: string[];
}

/** Detect secret-like keys (values are never returned — only key names). */
export function detectSecrets(obj: unknown, depth = 0): SecretDetection {
  const out: SecretDetection = { secret_fields_detected: [], seller_secrets_detected: [], provider_secrets_detected: [], plaintext_must_not_persist: [] };
  if (depth > 32 || obj === null || obj === undefined) return out;
  if (Array.isArray(obj)) {
    for (const item of obj) {
      const r = detectSecrets(item, depth + 1);
      mergeDetection(out, r);
    }
    return out;
  }
  if (typeof obj === "object") {
    for (const [key, value] of Object.entries(obj as Record<string, unknown>)) {
      const lower = key.toLowerCase();
      if (SELLER_SECRET_FIELDS.some((s) => lower.includes(s))) {
        out.secret_fields_detected.push(key);
        out.seller_secrets_detected.push(key);
        out.plaintext_must_not_persist.push(key);
      } else if (PROVIDER_SECRET_FIELDS.some((s) => lower.includes(s))) {
        out.secret_fields_detected.push(key);
        out.provider_secrets_detected.push(key);
      }
      const r = detectSecrets(value, depth + 1);
      mergeDetection(out, r);
    }
  }
  return out;
}

function mergeDetection(target: SecretDetection, src: SecretDetection): void {
  for (const key of ["secret_fields_detected", "seller_secrets_detected", "provider_secrets_detected", "plaintext_must_not_persist"] as const) {
    for (const v of src[key]) {
      if (!target[key].includes(v)) target[key].push(v);
    }
  }
}

/** Decide what to do with a detected provider secret. */
export function providerSecretDecision(options: LegacyImportOptions, providerKey: string): { action: "skip" | "store_ref"; ref?: string } {
  if (options.import_provider_secret === true) {
    return { action: "store_ref", ref: "cred_" + providerKey.toLowerCase().replace(/[^a-z0-9]+/g, "_").replace(/^_+|_+$/g, "") };
  }
  return { action: "skip" };
}

File: services/ai-worker/src/fastwork_ai_worker/legacy_import/secret_policy.py
"""Legacy secret policy (M11, clean-room). Worker-side mirror: detects secret
field names (never values); provider secrets SKIP by default and move to a
credential_ref only on explicit consent. Seller auth material is never imported.
"""
from __future__ import annotations

from typing import Any, Dict, List

SELLER_SECRET_FIELDS = ("cookie", "cookies", "session", "session_token", "token", "password", "passwd", "auth", "authorization", "device_fingerprint", "license", "login_state", "credential")
PROVIDER_SECRET_FIELDS = ("api_key", "apikey", "secret", "secret_key", "access_key", "app_secret")


def detect_secrets(obj: Any, depth: int = 0) -> Dict[str, List[str]]:
    out: Dict[str, List[str]] = {"secret_fields_detected": [], "seller_secrets_detected": [], "provider_secrets_detected": []}
    if depth > 32 or obj is None:
        return out
    if isinstance(obj, list):
        for item in obj:
            r = detect_secrets(item, depth + 1)
            for k in out:
                for v in r[k]:
                    if v not in out[k]:
                        out[k].append(v)
        return out
    if isinstance(obj, dict):
        for key, value in obj.items():
            lower = str(key).lower()
            if any(s in lower for s in SELLER_SECRET_FIELDS):
                out["secret_fields_detected"].append(str(key))
                out["seller_secrets_detected"].append(str(key))
            elif any(s in lower for s in PROVIDER_SECRET_FIELDS):
                out["secret_fields_detected"].append(str(key))
                out["provider_secrets_detected"].append(str(key))
            r = detect_secrets(value, depth + 1)
            for k in out:
                for v in r[k]:
                    if v not in out[k]:
                        out[k].append(v)
    return out


def provider_secret_decision(options: Dict[str, Any], provider_key: str) -> Dict[str, Any]:
    if options.get("import_provider_secret") is True:
        return {"action": "store_ref", "ref": "cred_" + provider_key.lower().replace(" ", "_")}
    return {"action": "skip"}
